framework/src/base/NonlinearSystemBase.C

```diff
@@ -2308,6 +2314,69 @@ NonlinearSystemBase::residualGhosted()
 }
 
 void
+NonlinearSystemBase::updateJacobianSparsity(SparseMatrix<Number> & jacobian)
+{
+  // This is computed in libMesh according to the mesh
+  const std::vector<dof_id_type> & basic_n_nz = dofMap().get_n_nz();
+  const std::vector<dof_id_type> & basic_n_oz = dofMap().get_n_oz();
+
+  std::vector<dof_id_type> n_nz(basic_n_nz.begin(), basic_n_nz.end());
+  std::vector<dof_id_type> n_oz(basic_n_oz.begin(), basic_n_oz.end());
+
+  _fe_problem.updateGeomSearch();
+
+  std::map<dof_id_type, std::vector<dof_id_type>> graph;
+
+  findImplicitGeometricCouplingEntries(_fe_problem.geomSearchData(), graph);
+
+  if (_fe_problem.getDisplacedProblem())
+    findImplicitGeometricCouplingEntries(_fe_problem.getDisplacedProblem()->geomSearchData(),
+                                         graph);
+
+  const dof_id_type first_dof_on_proc = dofMap().first_dof(processor_id());
+  const dof_id_type end_dof_on_proc = dofMap().end_dof(processor_id());
+
+  // The total number of dofs on and off processor
+  const dof_id_type n_dofs_on_proc = dofMap().n_local_dofs();
+  const dof_id_type n_dofs_not_on_proc = dofMap().n_dofs() - dofMap().n_local_dofs();
+
+  for (const auto & git : graph)
+  {
+    dof_id_type dof = git.first;
+    dof_id_type local_dof = dof - first_dof_on_proc;
+
+    if (dof < first_dof_on_proc || dof >= end_dof_on_proc)
+      continue;
+
+    const std::vector<dof_id_type> & row = git.second;
+
+    for (const auto & coupled_dof : row)
+    {
+      if (coupled_dof < first_dof_on_proc || coupled_dof >= end_dof_on_proc)
+      {
+        if (n_oz[local_dof] < n_dofs_not_on_proc)
+          n_oz[local_dof]++;
+      }
+      else
+      {
+        if (n_nz[local_dof] < n_dofs_on_proc)
+          n_nz[local_dof]++;
+      }
+    }
+  }
+#ifdef LIBMESH_HAVE_PETSC
+#if !PETSC_VERSION_LESS_THAN(3, 7, 2)
+  MatXAIJSetPreallocation(static_cast<PetscMatrix<Number> &>(jacobian).mat(),
+                          dofMap().block_size(),
+                          numeric_petsc_cast(n_nz.empty() ? nullptr : &n_nz[0]),
+                          numeric_petsc_cast(n_oz.empty() ? nullptr : &n_oz[0]),
+                          nullptr,
+                          nullptr);
```

Comment:
PETSc does not take your updated `n_nz` and `n_oz` in your implementation. I am still thinking about the interface.  When the interface is addressed, I will assist you with  your case.